application/src/ext/java/org/opentripplanner/ext/carpooling/service/DefaultCarpoolingService.java

```diff
@@ -0,0 +1,233 @@
+    WgsCoordinate passengerDropoff = new WgsCoordinate(request.to().getCoordinate());
+    var passengerDepartureTime = request.dateTime();
+    var searchWindow = request.searchWindow() == null
+      ? Duration.ofMinutes(30)
+      : request.searchWindow();
+
+    LOG.debug(
+      "Finding carpool itineraries from {} to {} at {}",
+      passengerPickup,
+      passengerDropoff,
+      passengerDepartureTime
+    );
+
+    var allTrips = repository.getCarpoolTrips();
+    LOG.debug("Repository contains {} carpool trips", allTrips.size());
+
+    var candidateTrips = allTrips
+      .stream()
+      .filter(trip ->
+        preFilters.accepts(
+          trip,
+          passengerPickup,
+          passengerDropoff,
+          passengerDepartureTime,
+          searchWindow
+        )
+      )
+      .toList();
+
+    LOG.debug(
+      "{} trips passed pre-filters ({} rejected)",
+      candidateTrips.size(),
+      allTrips.size() - candidateTrips.size()
+    );
+
+    if (candidateTrips.isEmpty()) {
+      return List.of();
+    }
+
+    var router = new CarpoolStreetRouter(
+      graph,
+      vertexLinker,
+      streetLimitationParametersService,
+      request
+    );
+    var insertionEvaluator = new InsertionEvaluator(router::route, delayConstraints);
+
+    // Find optimal insertions for remaining trips
+    var insertionCandidates = candidateTrips
+      .stream()
+      .map(trip -> {
+        List<InsertionPosition> viablePositions = positionFinder.findViablePositions(
+          trip,
+          passengerPickup,
+          passengerDropoff
+        );
+
+        if (viablePositions.isEmpty()) {
+          LOG.debug("No viable positions found for trip {} (avoided all routing!)", trip.getId());
+          return null;
+        }
+
+        LOG.debug(
+          "{} viable positions found for trip {}, evaluating with routing",
+          viablePositions.size(),
+          trip.getId()
+        );
+
+        // Evaluate only viable positions with expensive routing
+        return insertionEvaluator.findBestInsertion(
+          trip,
+          viablePositions,
+          passengerPickup,
+          passengerDropoff
+        );
+      })
+      .filter(Objects::nonNull)
+      .sorted(Comparator.comparing(InsertionCandidate::additionalDuration))
+      .limit(DEFAULT_MAX_CARPOOL_RESULTS)
+      .toList();
+
+    LOG.debug("Found {} viable insertion candidates", insertionCandidates.size());
+
+    // Map to itineraries
+    var itineraries = insertionCandidates
+      .stream()
+      .map(candidate -> itineraryMapper.toItinerary(request, candidate))
+      .filter(Objects::nonNull)
+      .toList();
+
+    LOG.info("Returning {} carpool itineraries", itineraries.size());
+    return itineraries;
+  }
+
+  private void validateRequest(RouteRequest request) throws RoutingValidationException {
+    if (
+      Objects.requireNonNull(request.from()).lat == null ||
```

Comment:
Do the requierd not null before the if´s, not inside the if checks - it makes it much harder to read.
```
      Objects.requireNonNull(request.from())
      Objects.requireNonNull(request.to())
```